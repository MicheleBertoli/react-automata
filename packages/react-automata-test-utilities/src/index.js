/* eslint-disable import/prefer-default-export */

export { default as testStateMachine } from './testStateMachine'
