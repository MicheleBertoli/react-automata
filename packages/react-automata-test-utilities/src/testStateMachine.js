import idx from 'idx'
import React from 'react'
import TestRenderer from 'react-test-renderer'
import { getShortestPaths } from 'xstate/lib/graph'
import { stringify } from 'react-automata-utilities'

const testStateMachine = (Component, options = {}) => {
  const paths = getShortestPaths(Component.machine, options.extendedState)

  Object.keys(paths).forEach(key => {
    const initialData = idx(options, _ => _.fixtures.initialData)
    const renderer = TestRenderer.create(
      <Component initialData={initialData} />
    )
    const instance = renderer.getInstance()

    paths[key].forEach(({ event, state }) => {
      const fixtures = idx(options, _ => _.fixtures[state][event])

      instance.handleTransition(event, fixtures)
    })

    const machineState =
      instance.state.machineState.toString() ||
      stringify(instance.state.machineState.value).join(',')

    expect(renderer.toJSON()).toMatchSnapshot(machineState)
  })
}

export default testStateMachine
