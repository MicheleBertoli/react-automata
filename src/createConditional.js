import idx from 'idx'
import memoize from 'memoize-one'
import PropTypes from 'prop-types'
import React from 'react'
import Context from './context'
import { DEFAULT_CHANNEL, getPatterns, matches } from './utils'

const createConditional = (displayName, contextField) => {
  class Conditional extends React.Component {
    componentDidMount() {
      if (this.isVisible && this.props.onShow) {
        this.props.onShow()
      }
    }

    componentDidUpdate() {
      if (!this.wasVisible && this.isVisible && this.props.onShow) {
        this.props.onShow()
      }

      if (this.wasVisible && !this.isVisible && this.props.onHide) {
        this.props.onHide()
      }
    }

    getPatterns = memoize(getPatterns)

    matches = memoize(matches)

    render() {
      this.wasVisible = this.isVisible
      const patterns = this.getPatterns(this.props.is)
      this.isVisible = this.matches(patterns, this.props.value)

      if (this.props.render) {
        return this.props.render(this.isVisible)
      }

      return this.isVisible ? this.props.children : null
    }
  }

  const Container = props => (
    <Context.Consumer>
      {context => (
        <Conditional
          {...props}
          value={idx(
            context,
            _ => _[props.channel || DEFAULT_CHANNEL][contextField]
          )}
        />
      )}
    </Context.Consumer>
  )

  Container.defaultProps = {
    children: null,
  }

  Container.displayName = displayName

  Container.propTypes = {
    channel: PropTypes.string,
    children: PropTypes.node,
    is: PropTypes.oneOfType([
      PropTypes.arrayOf(PropTypes.string),
      PropTypes.string,
    ]),
    render: PropTypes.func,
    onHide: PropTypes.func,
    onShow: PropTypes.func,
  }

  return Container
}

export default createConditional
