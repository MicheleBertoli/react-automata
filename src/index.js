export { default as Action } from './Action'
export { default as State } from './State'
export { default as testStateMachine } from './testStateMachine'
export { default as withStateMachine } from './withStateMachine'
