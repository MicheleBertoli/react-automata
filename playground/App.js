import React from 'react'
import { hot } from 'react-hot-loader'
import { Action, withStateMachine } from '../src'

const statechart = {
  initial: 'idle',
  states: {
    idle: {
      on: {
        FETCH: 'fetching',
      },
      onEntry: 'showButton',
    },
    fetching: {
      on: {
        SUCCESS: 'success',
        ERROR: 'error',
      },
      onEntry: 'fetchGists',
    },
    success: {
      onEntry: 'showGists',
    },
    error: {
      on: {
        FETCH: 'fetching',
      },
      onEntry: 'showError',
    },
  },
}

class App extends React.Component {
  fetchGists() {
    fetch('https://api.github.com/users/gaearon/gists')
      .then(response => response.json())
      .then(gists => this.props.transition('SUCCESS', { gists }))
      .catch(() => this.props.transition('ERROR'))
  }

  handleClick = () => {
    this.props.transition('FETCH')
  }

  render() {
    return (
      <div>
        <h1>Actions</h1>
        <Action is="showButton">
          <button onClick={this.handleClick}>Fetch</button>
        </Action>
        <Action is="fetchGists">Loading...</Action>
        <Action is="showGists">
          <ul>
            {this.props.gists.filter(gist => gist.description).map(gist => (
              <li key={gist.id}>{gist.description}</li>
            ))}
          </ul>
        </Action>
        <Action is="showError">
          <button onClick={this.handleClick}>Retry</button>
          Oh, snap!
        </Action>
      </div>
    )
  }
}

App.defaultProps = {
  gists: [],
}

const options = {
  devTools: true,
}

const StateMachine = withStateMachine(statechart, options)(App)

export default hot(module)(StateMachine)
