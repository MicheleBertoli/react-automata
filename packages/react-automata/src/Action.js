import createConditional from './createConditional'

export default createConditional('Action', 'actions')
