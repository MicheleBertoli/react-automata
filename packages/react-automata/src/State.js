import createConditional from './createConditional'

export default createConditional('State', 'machineState')
