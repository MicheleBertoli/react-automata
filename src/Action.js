import PropTypes from 'prop-types'
import createConditional from './createConditional'

const displayName = 'Action'

const propTypes = {
  hide: PropTypes.oneOfType([
    PropTypes.arrayOf(PropTypes.string),
    PropTypes.string,
  ]),
  show: PropTypes.oneOfType([
    PropTypes.arrayOf(PropTypes.string),
    PropTypes.string,
  ]),
}

const matches = (value, actions) =>
  Array.isArray(value)
    ? actions.some(action => value.includes(action))
    : actions.includes(value)

export const shouldShow = (props, context) =>
  matches(props.show, context.actions)

export const shouldHide = (props, context) =>
  props.hide
    ? matches(props.hide, context.actions)
    : !matches(props.show, context.actions)

export default createConditional({
  displayName,
  propTypes: process.env.NODE_ENV !== 'production' ? propTypes : null,
  shouldShow,
  shouldHide,
})
