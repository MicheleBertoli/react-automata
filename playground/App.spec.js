import { testStateMachine } from '../src'
import App from './App'

global.fetch = jest.fn(() => new Promise(resolve => resolve()))

const fixtures = {
  fetching: {
    SUCCESS: {
      gists: [
        {
          id: 'ID1',
          description: 'GIST1',
        },
        {
          id: 'ID2',
          description: 'GIST2',
        },
      ],
    },
  },
}

test('it works', () => {
  testStateMachine(App, { fixtures })
})
