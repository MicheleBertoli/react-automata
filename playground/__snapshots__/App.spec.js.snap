// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`it works: error 1`] = `
<div>
  <h1>
    Actions
  </h1>
  <button
    onClick={[Function]}
  >
    Retry
  </button>
  Oh, snap!
</div>
`;

exports[`it works: fetching 1`] = `
<div>
  <h1>
    Actions
  </h1>
  Loading...
</div>
`;

exports[`it works: idle 1`] = `
<div>
  <h1>
    Actions
  </h1>
  <button
    onClick={[Function]}
  >
    Fetch
  </button>
</div>
`;

exports[`it works: success 1`] = `
<div>
  <h1>
    Actions
  </h1>
  <ul>
    <li>
      GIST1
    </li>
    <li>
      GIST2
    </li>
  </ul>
</div>
`;
